docs_src/design/adr/security/0026-authentication.md

```diff
@@ -0,0 +1,296 @@
+
+Behind the proxy, there are two major changes:
+
+1. Every EdgeX service, when security is enabled,
+   requires a JWT be passed as part of the HTTP request
+   that is validated using Vault's token introspection endpoint.
+
+2. Every EdgeX service, when security is enabled,
+   uses a Vault-supplied JWT to authenticate
+   outgoing calls to peer EdgeX services.
+   For microservice chaining scenarios,
+   the original caller's identity can be passed through
+   in most cases.
+
+The new TO-BE architecture is diagrammed in the following figure:
+
+![TO-BE Architecture](0026-to-be.jpg)
+
+The other architecturally signficant change to the proposal
+is the use Vault's identity features.
+Today, every EdgeX service is assigned a token with defined
+access to the EdgeX secret store.
+With this proposal, every service, and every external user,
+will will have an assigned Vault identity.
+The implication of this change is that it is now possible
+to configure Vault authentication engines against
+Vault identities.
+For example:
+
+* An external user identity could be authenticated by
+  an external service, such as [Auth0](https://auth0.com).
+  Alternatively, username/password or AppRole authentication
+  could be used if an external source of identity is not available.
+  This is viewed as beneficial, as downstream EdgeX implementations
+  are already building their own similar integrations.
+
+* An internal service identity could be authenticated by
+  a Kubernetes service account token.  This would eliminate
+  the requirement to pre-distribute Vault tokens to services
+  via a shared filesystem volume, simplyfing Kubenetes-based
+  deployments of EdgeX.
+
+As an added bonus,
+Vault supports longer JWT key sizes than the Kong JWT plugin.
+
+
+## Alternatives
+
+### Using Kong to Mediate EdgeX Microservice Interactions
+
+One approach that is seen in some microservice architectures
+is to force all communication between microservices to go
+through the external API gateway.
+The main barrier to this approach is lack of
+proces-level traffic shaping in bare-metal environments.
+Additionally, external address of the API gateway
+may not be known to internal code,
+increasing implementation difficulty for the programmer.
+
+### mTLS Everywhere
+
+One straightforward approach would be to use
+use mutual-auth TLS everywhere and eliminate
+the reverse proxy entirely.
+Each service would necessarily have to be exposed
+on its own port in order to process the client certificate.
+A large number of exposed ports makes
+security validation more difficult.
+
+The primary problem with mutual-auth TLS
+is key and certificate management.
+Specifically,
+[NIST SP 800-57 part 1](https://csrc.nist.gov/publications/detail/sp/800-57-part-1/rev-5/final)
+recommends that authentication keys
+should be rotated every 2 years as a minimum.
+This is a potential issue for long-running
+(for multiple years) processes.
```

Comment:
Yes - but wouldn't many organizations expect mTLS to be an option?  Does this new architecture/solution make it impossible (or more difficult) to put in - provided they also solve the key management/updates/etc.?

Reply:
There are a few aspects to this:

(1) Layering one-way TLS on top of JWT authentication is sound practice to project the JWT on the wire.

(2) It is entirely possible to stick a JWT inside an mTLS (or plain TLS) connection.  Client certificate authentication is stronger because it isn't subject to man-in-the-middle attacks.  But if you have both, you have to decide the responsibility for each.  What information is carried in the JWT and what is carried in the certificate?  Layering approach is possible:  mTLS gives you permission to talk to me, JWT tells me who you are.  (For example, mTLS could tell you that NGINX sent the packet, but the JWT says who you are.  Also as an FYI, NGINX supports upstream TLS but Kong does not.)   If one decides mTLS does both who you are and are allowed to talk, in which case JWT is redundant.

(3) mTLS is also hard on developers if they need to debug something because it is almost impossible to observe encrypted network traffic over the wire.  Authentication is wrapped up with encryption, so if you disable encryption for observability, everything stops working because all the authentication info vanishes along with it.

On the face of it, "does using JWT make it more difficult to use mTLS", the answer is "no".

Reply:
A lot of this discussion was moved to the UCR.